llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -8404,6 +8404,83 @@ SDValue RISCVTargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
         IntNo == Intrinsic::riscv_zip ? RISCVISD::ZIP : RISCVISD::UNZIP;
     return DAG.getNode(Opc, DL, XLenVT, Op.getOperand(1));
   }
+  case Intrinsic::riscv_mopr: {
+    unsigned mopr_id = cast<ConstantSDNode>(Op.getOperand(2))->getZExtValue();
+    switch (mopr_id) {
+#define RISCV_MOPR_64_CASE(MOPR_ID, OPCODE)                                    \
+  case MOPR_ID: {                                                              \
+    if (RV64LegalI32 && Subtarget.is64Bit() &&                                 \
+        Op.getValueType() == MVT::i32) {                                       \
+      SDValue NewOp =                                                          \
+          DAG.getNode(ISD::ANY_EXTEND, DL, MVT::i64, Op.getOperand(1));        \
+      SDValue Res = DAG.getNode(OPCODE, DL, MVT::i64, NewOp);                  \
+      return DAG.getNode(ISD::TRUNCATE, DL, MVT::i32, Res);                    \
+    }                                                                          \
+    return DAG.getNode(OPCODE, DL, XLenVT, Op.getOperand(1));                  \
+  }
+      RISCV_MOPR_64_CASE(0, RISCVISD::MOPR0)
```

Comment:
Can we add `RISCVISD::MOPRR` and `RISCVISD::MOPR` and do the immediate matching in tblgen patterns.

Reply:
Are you going to make this change?

Reply:
I have tried to change it as you asked. Unfortunately, I had some trouble with it. I am not very competent in tblgen syntax and could not get corresponding instruction which depends on the third operand. If you can give an example it will be nice.

Reply:
I took this patch and modified it here https://github.com/topperc/llvm-project/commit/401f1e6a42f7760bc5e09426a74e69df8c37de72

Reply:
I guess @ln8-8 tried to append the third argument to the instruction name and avoid defining 8/32 patterns for mapr/moprr instructions. 
Did something like this

```
def : Pat<(XLenVT (riscv_mopr GPR:$rs1, uimm5:$imm)),
            (!cast<Instruction>("MOPR"#!repr($imm->getSExtValue())) GPR:$rs1)>;
```

Anyway, Thank you.